sdk/webpubsub/Microsoft.Azure.WebJobs.Extensions.WebPubSub/src/Trigger/WebPubSubTriggerDispatcher.cs

```diff
@@ -0,0 +1,336 @@
+                    Subprotocols = subprotocols,
+                    ClientCertificates = certificates,
+                    Reason = reason,
+                    TaskCompletionSource = tcs
+                };
+                await executor.Executor.TryExecuteAsync(new TriggeredFunctionData
+                {
+                    TriggerValue = triggerEvent
+                }, token).ConfigureAwait(false);
+
+                // After function processed, return on-hold event reponses.
+                if (requestType == RequestType.Connect || requestType == RequestType.User)
+                {
+                    try
+                    {
+                        using (token.Register(() => tcs.TrySetCanceled()))
+                        {
+                            var response = await tcs.Task.ConfigureAwait(false);
+
+                            // Skip no returns
+                            if (response != null)
+                            {
+                                var validResponse = BuildValidResponse(response, requestType);
+
+                                if (validResponse != null)
+                                {
+                                    return validResponse;
+                                }
+                                _logger.LogWarning($"Invalid response type {response.GetType()} regarding current request: {requestType}");
+                            }
+                        }
+                    }
+                    catch (Exception ex)
+                    {
+                        var error = new ErrorResponse(WebPubSubErrorCode.ServerError, ex.Message);
+                        return Utilities.BuildErrorResponse(error);
+                    }
+                }
+
+                return new HttpResponseMessage(HttpStatusCode.OK);
+            }
+            // No function map to current request
+            return new HttpResponseMessage(HttpStatusCode.NotFound);
+        }
+
+        private static bool TryParseRequest(HttpRequestMessage request, out ConnectionContext context)
+        {
+            // ConnectionId is required in upstream request, and method is POST.
+            if (!request.Headers.Contains(Constants.Headers.CloudEvents.ConnectionId)
+                || request.Method != HttpMethod.Post)
+            {
+                context = null;
+                return false;
+            }
+
+            context = new ConnectionContext();
+            try
+            {
+                context.ConnectionId = request.Headers.GetValues(Constants.Headers.CloudEvents.ConnectionId).FirstOrDefault();
+                context.Hub = request.Headers.GetValues(Constants.Headers.CloudEvents.Hub).FirstOrDefault();
+                context.EventType = Utilities.GetEventType(request.Headers.GetValues(Constants.Headers.CloudEvents.Type).FirstOrDefault());
+                context.EventName = request.Headers.GetValues(Constants.Headers.CloudEvents.EventName).FirstOrDefault();
```

Comment:
Would it make sense to use SingleOrDefault for these if the contract requires at most 1 value?